swarm/src/behaviour.rs

```diff
@@ -152,6 +152,11 @@ type THandlerInEvent<THandler> =
 /// struct members are delegated to [`NetworkBehaviourEventProcess`] implementations. Those must be
 /// provided by the user on the type that [`NetworkBehaviour`] is derived on.
 ///
+/// Because these implementations are granted exclusive access to the [`NetworkBehaviour`],
+/// [blocking code](https://ryhl.io/blog/async-what-is-blocking/) in these implementations will
+/// stop the entire swarm from processing new events, since the swarm cannot progress without also
+/// having access exclusive access to the NetworkBehaviour.
```

Comment:
How about adding something like

> A better alternative is to spawn blocking or asynchronous tasks on a (separate) executor, which may be provided to the behaviour’s constructor.

It would be nice if the network pool’s `executor` were available to (sub)behaviours for spawning tasks. @mxinden what do you think?

Reply:
Added it, that's a good suggestion!

Reply:
And I think the suggestion of providing the async executor in the networkbehaviour implementation would be great, I went with the dirty solution of rewriting the function I needed to call with blocking code. In the future, I'll probably try to provide access to the executor by including it in the NetworkBehaviour struct as you wrote in your suggested addition.

Reply:
> It would be nice if the network pool’s executor were available to (sub)behaviours for spawning tasks. @mxinden what do you think?

Indeed, that would be an option and would make things a bit more consistent instead of requiring users to pass yet another executor to the behaivour.

What would be your use-case @rkuhn?

In my eyes, rust-libp2p greatly benefits of being designed around a single synchronous hierarchical state machine (`Swarm` + `NetworkBehaviour`) + one hierarchical state machine per connection (managed via `Pool`).

Reply:
My use-case is a converter between Rust structs (from a channel) and substream frames (going into AsyncWrite). Polling the whole swarm (easily a hierarchy of a dozen or more state machines, most of them uninterested), for passing a single message through is wasteful. Or sending a notification from a behaviour to client code via a bounded channel — unbounded channels creep me out, they’re an OOM waiting to happen. In the second case the computational overhead may in some cases be tolerable, but the coding overhead is burdensome — FuturesUnordered galore.

Reply:
> unbounded channels creep me out, they’re an OOM waiting to happen.

:+1: 

> My use-case is a converter between Rust structs (from a channel) and substream frames (going into AsyncWrite). Polling the whole swarm (easily a hierarchy of a dozen or more state machines, most of them uninterested), for passing a single message through is wasteful.

Does this has to happen in the main task (`Swarm` and `NetworkBehaviour`) or can this happen in the connection task (`ProtocolsHandler`)?